File: backend-fastapi/app/services/auth_otp_service.py
import secrets
from datetime import datetime, timedelta

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.config import settings
from app.core.security import hash_password
from app.models.otp import EmailOtp
from app.models.users import Users
from app.schemas.auth import OtpDispatchResponse
from app.services import email_service

RESET = "RESET"
VERIFY = "VERIFY"


def _norm(email: str | None) -> str:
    return (email or "").strip().lower()


def _generate_otp() -> str:
    return f"{secrets.randbelow(1_000_000):06d}"


def _create_otp(db: Session, email: str, purpose: str) -> str:
    otp = EmailOtp(
        email=email,
        otp=_generate_otp(),
        purpose=purpose,
        expires_at=datetime.now() + timedelta(minutes=settings.otp_expiry_minutes),
        used=False,
        created_at=datetime.now(),
    )
    db.add(otp)
    db.commit()
    return otp.otp


def _dispatch(otp: str) -> OtpDispatchResponse:
    delivered = email_service.is_configured()
    return OtpDispatchResponse(delivered=delivered, devOtp=None if delivered else otp)


def forgot_password(db: Session, raw_email: str | None) -> OtpDispatchResponse:
    email = _norm(raw_email)
    if not email:
        raise ValueError("Email is required")

    user = db.scalar(select(Users).where(Users.email == email))
    if user is None:
        # Pretend success; nothing sent (don't reveal whether the email exists).
        return OtpDispatchResponse(delivered=email_service.is_configured(), devOtp=None)

    otp = _create_otp(db, email, RESET)
    email_service.send_otp(email, otp, "password reset")
    return _dispatch(otp)


def send_email_verification(db: Session, raw_email: str | None) -> OtpDispatchResponse:
    email = _norm(raw_email)
    user = db.scalar(select(Users).where(Users.email == email))
    if user is None:
        raise ValueError("No account found for this email")

    otp = _create_otp(db, email, VERIFY)
    email_service.send_otp(email, otp, "email verification")
    return _dispatch(otp)


def _find_latest_unused(db: Session, email: str, purpose: str) -> EmailOtp | None:
    return db.scalar(
        select(EmailOtp)
        .where(EmailOtp.email == _norm(email), EmailOtp.purpose == purpose, EmailOtp.used.is_(False))
        .order_by(EmailOtp.id.desc())
        .limit(1)
    )


def _is_valid(db: Session, email: str, otp: str | None, purpose: str) -> bool:
    if not otp or not otp.strip():
        return False
    record = _find_latest_unused(db, email, purpose)
    if record is None:
        return False
    if record.expires_at <= datetime.now():
        return False
    return record.otp == otp.strip()


def verify_otp(db: Session, email: str | None, otp: str | None, purpose: str | None) -> bool:
    p = VERIFY if (purpose or "").upper() == VERIFY else RESET
    return _is_valid(db, email or "", otp, p)


def reset_password(db: Session, email: str | None, otp: str | None, new_password: str | None) -> None:
    if not new_password or len(new_password) < 6:
        raise ValueError("Password must be at least 6 characters")

    record = _find_latest_unused(db, email or "", RESET)
    if (
        record is None
        or record.expires_at <= datetime.now()
        or record.otp != (otp.strip() if otp else None)
    ):
        raise ValueError("Invalid or expired code")

    user = db.scalar(select(Users).where(Users.email == _norm(email)))
    if user is None:
        raise ValueError("Account not found")

    user.password = hash_password(new_password)
    record.used = True
    db.commit()


def verify_email(db: Session, email: str | None, otp: str | None) -> None:
    record = _find_latest_unused(db, email or "", VERIFY)
    if (
        record is None
        or record.expires_at <= datetime.now()
        or record.otp != (otp.strip() if otp else None)
    ):
        raise ValueError("Invalid or expired code")

    user = db.scalar(select(Users).where(Users.email == _norm(email)))
    if user is None:
        raise ValueError("Account not found")

    user.email_verified = True
    record.used = True
    db.commit()

File: backend/src/main/java/com/fitnexus/backend/service/AuthOtpService.java
package com.fitnexus.backend.service;

import com.fitnexus.backend.dto.OtpDispatchResponse;
import com.fitnexus.backend.entity.EmailOtp;
import com.fitnexus.backend.entity.Users;
import com.fitnexus.backend.repository.EmailOtpRepository;
import com.fitnexus.backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
@Transactional
public class AuthOtpService {
    private final UserRepository userRepository;
    private final EmailOtpRepository emailOtpRepository;
    private final PasswordEncoder passwordEncoder;
    private final EmailService emailService;

    private final SecureRandom random = new SecureRandom();

    @Value("${app.otp.expiry-minutes:10}")
    private int expiryMinutes;

    private static final String RESET = "RESET";
    private static final String VERIFY = "VERIFY";

    private String norm(String email) {
        return email == null ? "" : email.trim().toLowerCase();
    }

    private String generateOtp() {
        return String.format("%06d", random.nextInt(1_000_000));
    }

    private String createOtp(String email, String purpose) {
        EmailOtp otp = new EmailOtp();
        otp.setEmail(email);
        otp.setOtp(generateOtp());
        otp.setPurpose(purpose);
        otp.setExpiresAt(LocalDateTime.now().plusMinutes(expiryMinutes));
        otp.setUsed(false);
        otp.setCreatedAt(LocalDateTime.now());
        emailOtpRepository.save(otp);
        return otp.getOtp();
    }

    private OtpDispatchResponse dispatch(String otp) {
        boolean delivered = emailService.isConfigured();
        return new OtpDispatchResponse(delivered, delivered ? null : otp);
    }

    /** Always returns success (doesn't reveal whether the email exists). */
    public OtpDispatchResponse forgotPassword(String rawEmail) {
        String email = norm(rawEmail);
        if (email.isBlank()) {
            throw new IllegalArgumentException("Email is required");
        }
        if (userRepository.findByEmail(email).isEmpty()) {
            // Pretend success; nothing sent (don't reveal whether the email exists).
            return new OtpDispatchResponse(emailService.isConfigured(), null);
        }
        String otp = createOtp(email, RESET);
        emailService.sendOtp(email, otp, "password reset");
        return dispatch(otp);
    }

    public OtpDispatchResponse sendEmailVerification(String rawEmail) {
        String email = norm(rawEmail);
        // Validate the account exists (throws if not); the entity itself isn't needed here.
        userRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("No account found for this email"));
        String otp = createOtp(email, VERIFY);
        emailService.sendOtp(email, otp, "email verification");
        return dispatch(otp);
    }

    private boolean isValid(String email, String otp, String purpose) {
        if (otp == null || otp.isBlank()) return false;
        return emailOtpRepository.findTopByEmailAndPurposeAndUsedFalseOrderByIdDesc(norm(email), purpose)
                .filter(o -> o.getExpiresAt().isAfter(LocalDateTime.now()))
                .filter(o -> o.getOtp().equals(otp.trim()))
                .isPresent();
    }

    /** Non-consuming validity check (used by the Verify Pin step). */
    public boolean verifyOtp(String email, String otp, String purpose) {
        String p = VERIFY.equalsIgnoreCase(purpose) ? VERIFY : RESET;
        return isValid(email, otp, p);
    }

    public void resetPassword(String email, String otp, String newPassword) {
        if (newPassword == null || newPassword.length() < 6) {
            throw new IllegalArgumentException("Password must be at least 6 characters");
        }
        EmailOtp record = emailOtpRepository
                .findTopByEmailAndPurposeAndUsedFalseOrderByIdDesc(norm(email), RESET)
                .filter(o -> o.getExpiresAt().isAfter(LocalDateTime.now()))
                .filter(o -> o.getOtp().equals(otp == null ? null : otp.trim()))
                .orElseThrow(() -> new IllegalArgumentException("Invalid or expired code"));

        Users user = userRepository.findByEmail(norm(email))
                .orElseThrow(() -> new IllegalArgumentException("Account not found"));
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);

        record.setUsed(true);
        emailOtpRepository.save(record);
    }

    public void verifyEmail(String email, String otp) {
        EmailOtp record = emailOtpRepository
                .findTopByEmailAndPurposeAndUsedFalseOrderByIdDesc(norm(email), VERIFY)
                .filter(o -> o.getExpiresAt().isAfter(LocalDateTime.now()))
                .filter(o -> o.getOtp().equals(otp == null ? null : otp.trim()))
                .orElseThrow(() -> new IllegalArgumentException("Invalid or expired code"));

        Users user = userRepository.findByEmail(norm(email))
                .orElseThrow(() -> new IllegalArgumentException("Account not found"));
        user.setEmailVerified(true);
        userRepository.save(user);

        record.setUsed(true);
        emailOtpRepository.save(record);
    }
}
